urlfd018a75f7_highlanderkev_StockExVirtuoso_tgz-pJ8-driver_Phase4J8/src/client/UserCommandService.java

```diff
@@ -0,0 +1,440 @@
+        if (verifyUser(userName, cId)) {
+            Order newOrder = new Order(userName, product, price, volume, side);
+            String orderId = ProductService.getInstance().submitOrder(newOrder);
+            return orderId;
+        } else {
+            return ("Invalid Order.");
+        }
+    }
+
+    /**
+     * This method will forward the provided information to the product service
+     * submitOrderCancel method.
+     *
+     * @param userName
+     * @param cId connection id
+     * @param product
+     * @param side
+     * @param orderId
+     * @throws Exception
+     */
+    public void submitOrderCancel(String userName, long cId, String product, BookSide side, String orderId) throws Exception {
+        if (verifyUser(userName, cId)) {
+            ProductService.getInstance().submitOrderCancel(product, side, orderId);
+        }
+    }
+
+    /**
+     * This method will create a quote object using the data passed in and will
+     * forward the quote to the product services submitQuote method.
+     *
+     * @param userName
+     * @param cId connection id
+     * @param product
+     * @param bPrice buy price
+     * @param bVolume buy volume
+     * @param sPrice sell price
+     * @param sVolume sell volume
+     * @throws Exception
+     */
+    public void submitQuote(String userName, long cId, String product, Price bPrice, int bVolume, Price sPrice, int sVolume) throws Exception {
+        if (verifyUser(userName, cId)) {
+            Quote newQuote = new Quote(userName, product, bPrice, bVolume, sPrice, sVolume);
+            ProductService.getInstance().submitQuote(newQuote);
+        }
+    }
+
+    /**
+     * This method will forward the provided data to the product services
+     * submitQuoteCancel method.
+     *
+     * @param userName
+     * @param cId connection id
+     * @param product
+     * @throws Exception
+     */
+    public void submitQuoteCancel(String userName, long cId, String product) throws Exception {
+        if (verifyUser(userName, cId)) {
+            ProductService.getInstance().submitQuoteCancel(userName, product);
+        }
+    }
+
+    /**
+     * This method will forward the subscription request to the current market
+     * publisher.
+     *
+     * @param userName
+     * @param cId connection id
+     * @param product
+     * @throws Exception
+     */
+    public void subscribeCurrentMarket(String userName, long cId, String product) throws Exception {
+        if (verifyUser(userName, cId)) {
+            if (connectedUsers.get(userName) == null) {
+                throw new UserNotConnectedException("UserNotConnectedException: User name passed to UserCommandService.subscribeCurrentMarket is not connected.");
+            } else {
+                User u = connectedUsers.get(userName);
+                CurrentMarketPublisher.getInstance().subscribe(u, product);
+            }
+        }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 283:290, 304:311, 325:332, 346:353, 367:374, 388:395, 409:416, 430:437.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.